extensions/lem-dashboard/lem-dashboard.lisp

```diff
@@ -0,0 +1,130 @@
+(defpackage :lem-dashboard
+  (:use :cl :lem)
+  (:export :open-dashboard
+           :*dashboard-mode-keymap*
+           :*dashboard-enable*
+           :move-to-recent-projects
+           :move-to-recent-files
+           :set-dashboard
+           :set-default-dashboard
+           :create-centered-string))
+
+(in-package :lem-dashboard)
+
+(defvar *dashboard-buffer-name* "*dashboard*")
+(defvar *dashboard-enable* t)
+
+(defun create-centered-string (str)
+  "Creates a 'centered' string by padding with space to fill the screen width halfway."
+  (let* ((padding (max 0 (floor (- (window-width (current-window)) (length str)) 2)))
+         (spaces (make-string padding :initial-element #\Space)))
+    (concatenate 'string spaces str)))
+
+(defclass dashboard-item ()
+  ((item-attribute 
+    :initarg :item-attribute 
+    :accessor item-attribute 
+    :initform 'document-text-attribute
+    :documentation "Attribute to use when drawing this item.")
+   (top-margin
+    :initarg :top-margin 
+    :accessor top-margin
+    :initform 0
+    :documentation "The amount of vertical space (lines) to apply before the item.")
+   (bottom-margin
+    :initarg :bottom-margin 
+    :accessor bottom-margin
+    :initform 1
+    :documentation "The amount of vertical space (lines) to apply after the item.")
+   (action 
+    :initarg :action 
+    :accessor action 
+    :initform nil
+    :documentation "Function to execute when <return> is pressed over this item."))
+  (:documentation "Base class for all dashboard items."))
+
+(defgeneric draw-dashboard-item (item point)
+  (:documentation "Called to draw the dashboard item.")
+  (:method :before ((item dashboard-item) point)
+    (dotimes (i (top-margin item))
+      (insert-character point #\Newline)))
+  (:method :after ((item dashboard-item) point)
+    (dotimes (i (bottom-margin item))
+      (insert-character point #\Newline))))
+
+(define-command dashboard-open-selected-item () ()
+  "Execute action on selected dashboard item."
+  (let* ((point (current-point))
+         (item (text-property-at point :dashboard-item)))
+    (when (and item (action item))
+      (funcall (action item)))))
+
+(defmethod draw-dashboard-item :around ((item dashboard-item) point)
+  "Inserts a :dashboard-item text property in between the starting and ending POINT, useful for tracking."
+  (let ((start (copy-point point :temporary)))
+    (call-next-method)
+    (let ((end (copy-point point :temporary)))
+      (put-text-property start end :dashboard-item item)
+      (delete-point start)
+      (delete-point end))))
+
+(defvar *dashboard-layout* nil
```

Comment:
variables and parameters should be together at the top of the file, before functions.

(pretty sure it's a general convention and the one we follow here)